pmd-ui/src/main/java/net/sourceforge/pmd/util/fxdesigner/XPathPanelController.java

```diff
@@ -112,9 +120,63 @@ public void initialize(URL location, ResourceBundle resources) {
                            // Reevaluate XPath anytime the expression or the XPath version changes
                            .or(xpathVersionProperty().changes())
                            .subscribe(tick -> parent.refreshXPathResults());
+
+        xpathExpressionArea.plainTextChanges()
+                           .map(m -> m.getInserted())
+                           .filter(t -> t.matches("[a-zA-Z/]"))
+                           .subscribe(e -> {
+                               if (e.matches("/")) {
+                                   autoComplete(xpathExpressionArea.getText().substring(xpathExpressionArea
+                                                                                            .getText().lastIndexOf("/", xpathExpressionArea.getText().lastIndexOf("/")),
+                                                                                        xpathExpressionArea.getCaretPosition() - 1));
+                               } else {
+                                   autoComplete(xpathExpressionArea.getText().substring(xpathExpressionArea
+                                                                                            .getText().lastIndexOf(("/"))));
+                               }
+                           });
+    }
+
+    private void autoComplete(String input) {
+        autoCompletePopup.getItems().clear();
+
+        XPathSuggestions xPathSuggestions = new XPathSuggestions(parent.getLanguageVersion().getTerseName());
+        List<String> suggestions = xPathSuggestions.getXPathSuggestions(input.replace("/", "").trim());
+
+        List<MenuItem> resultToDisplay = suggestions.stream().map(m -> new MenuItem(m)).limit(5).collect(Collectors
```

Comment:
You should probably create a class e.g. CompletionResult that encapsulates a suggestion and can be transformed to a menuitem. That class would know which segments of the suggestion were matched by the matching algorithm, in order to highlight them in a TextFlow. So the matching algorithm would return a list of CompletionResults